JsonPath Improvements

jsneedles opened the pull request:
I noticed #974 recently, and as a big user of JsonPath in C# -- I had to take a look.

There were a few very plain bugs that were present... But as I dug in more I realized that there was opportunity to make the JsonPath more closely aligned with the standard/rfc (https://www.rfc-editor.org/rfc/rfc9535) 

## Main Changes: 

- Added support for `Scan` filters for Index / Slicer / Multiple Index / Multiple Slicer
- Aligned `scan` recursive behavior with the spec (don't return self) and treat `.*` and `[*]` the same) 
- Added support for negative / from-the-end indices like `$.books[-1]` 
- Added support for `in` (case-insensitive) operator as well as `!` for NOT
- Added support for query filter groupings like `$[?(@.a && (@.b || @.c || @.z))]` - explicit parentheses threw errors previously
- Added support for parsing array literals (useful for `in`) so you can say `$[?(@.a in [1,2,3])]` 
- Fixes for a few matching bugs, especially for null handling 
- Supporting parsing quoted strings in single or double quotes
- Escaping check for unicode characters / code points

## New Tests
- I added a few new tests to validate my first fixes, but realized it'd be easier to lean on the JsonPath regression suite (https://cburgmer.github.io/json-path-comparison/).

There's an extensive set of cases that covers a wide range of parsing & evaluation scenarios, that this now successfully passes (mostly) 😄 .

There's a few remaining scenarios that fail, but wanted to get feedback / opinions on... Namely `large number indexers`. Right now we parse as `int` and it fails, if we switched that to `long`, the parsing would succeed, but I think trying to actually evaluate the index of an array that long won't work (and is pretty unrealistic anyhow). 

![image](https://github.com/user-attachments/assets/00e77a60-2bec-4063-8b32-ed70e5e39476)

I currently left the tests without a consensus that threw errors as `inconclusive`.

## Future

There's definitely room for optimization, but I ensured the benchmark results are in line with the results before my PR. 

- the `in` operator especially right now is naive, it could definitely be made more efficient!


Would love to gather some feedback and I'll tweak accordingly!

jsneedles commented:
@microsoft-github-policy-service agree

prvyk commented:
> There's a few remaining scenarios that fail, but wanted to get feedback / opinions on... Namely `large number indexers`. Right now we parse as `int` and it fails, if we switched that to `long`, the parsing would succeed, but I think trying to actually evaluate the index of an array that long won't work (and is pretty unrealistic anyhow).

I guess (without reading the code), there are some length check to handle indexes outside the array, and that ranges are capped at array length?

If so, I think the proper solution would be to read as long, but internally set a int sentinel value for too large values, and then let the length checks handle this. Note that the C# maximum is Array.MaxValue which is a tiny bit smaller than int.MaxValue so there's room for large sentinel values and no worry about collisions with an existing index or changing the variable type. Possibly no need to add special sentinel checks. The advantage is being slightly more compatible without complicating the implementation code.

There was a similar scenario with *SCAN cursor values and I made a similar choice to change the code to accept longs even though builtin Garnet objects would never get there as they are implemented now. (There was some extra motivation so it's not a complete parallel: It's possible to design custom objects and maybe someone might choose to use long cursors. Cursor values are opaque and allowed to pack more information than a single counter so the array/collection limits aren't definitive).

TalZaccai commented:
@Vijay-Nirmal Could you also take a look since this refers to your original PR?

jsneedles commented:
> > There's a few remaining scenarios that fail, but wanted to get feedback / opinions on... Namely `large number indexers`. Right now we parse as `int` and it fails, if we switched that to `long`, the parsing would succeed, but I think trying to actually evaluate the index of an array that long won't work (and is pretty unrealistic anyhow).
> 
> I guess (without reading the code), there are some array length checks to handle indexes outside the array, and that ranges are capped?
> 
> If so, I think the proper solution would be to read as long, but internally set a int sentinel value for too large values, and then let the length checks handle this. Note that the C# maximum is Array.MaxValue which is a tiny bit smaller than int.MaxValue so there's room for large sentinel values that do not need to worry about collisions with an existing index - or about changing the variable type. Possibly no need to add special sentinel checks. The advantage is being slightly more compatible without complicating the implementation code.
> 
> There was a similar scenario with *SCAN cursor values and I made a similar choice to change the code to accept longs even though builtin Garnet objects would never get there as they are implemented now. Though I just changed to long throughout (there was very little to change) and didn't need a sentinel. I suspect JsonPath is a bit more complicated though?
> 
> (There was some extra motivation so it's not a complete parallel: It's possible to design custom objects and maybe someone might choose to use long cursors. Cursor values are opaque and allowed to pack more information than a single counter so the array/collection limits aren't definitive).

Thanks! 

I think I got the gist of what you're suggesting is to just "cheat" 😄  -- so I went ahead and did that.

Since it's unrealistic for any array of these lengths to be had (and I never knew about `Array.MaxLength` so thx! )-- I just changed it to parse as a `long` and if it's outside the `Array.MaxLength` range, I just capped it at that. 

Let me know if you were thinking of something different.

prvyk commented:
> 
> Thanks!
> 
> I think I got the gist of what you're suggesting is to just "cheat" 😄 -- so I went ahead and did that.
> 
...
> 
> Let me know if you were thinking of something different.

Yep, I was thinking of what you did. It gives us the right result in any realistic case - a C# array can't contain more items.

I suspect that if it ever happens in real code it's because someone is using large values in a slice as a stand-in for infinite instead of doing an open start/end (maybe a jsonpath template is buried so deep they can't modify it easily, and can only supply integers to a function/API. 'Eh, let's give +-(2^53-1) and call it a day. We'll never have so many items anyway.').

TalZaccai reviewed (changes requested):
LGTM overall, added one comment. 
Probably need to run dotnet format, there are a bunch of extra whitespaces all over :)

jsneedles commented on modules/GarnetJSON/JSONPath/JsonPath.cs:
```diff
@@ -360,18 +366,17 @@ private PathFilter ParseArrayIndexer(char indexerCloseChar)
 
                         var indexer = _expression.AsSpan(start, length);
-                        int index = int.Parse(indexer, CultureInfo.InvariantCulture);
+                        int index = EnsureIndexWithinArrayBounds(long.Parse(indexer, CultureInfo.InvariantCulture));
```
Thanks! I went ahead and made that change.

jsneedles commented:
> LGTM overall, added one comment. Probably need to run dotnet format, there are a bunch of extra whitespaces all over :)

Thanks! 

I just ran `dotnet format` and don't see any changes 🤔   Also did `dotnet format --verify-no-changes` a bunch of times and no errors.

(I've not used `dotnet format` before, so maybe im missing some step/setup)

TalZaccai commented:
> > LGTM overall, added one comment. Probably need to run dotnet format, there are a bunch of extra whitespaces all over :)
> 
> Thanks!
> 
> I just ran `dotnet format` and don't see any changes 🤔 Also did `dotnet format --verify-no-changes` a bunch of times and no errors.
> 
> (I've not used `dotnet format` before, so maybe im missing some step/setup)

No worries, I ended up running code cleanup and that seems to have done the trick, hope that's ok.
